parcllabs/services/parcllabs_service.py

```diff
@@ -146,65 +258,65 @@ def retrieve(
+                data_container.extend(
+                    results if isinstance(results, list) else [results]
+                )
+            except NotFoundError:
+                # we don't want to kill the entire process if one of the chunks fails due to no data
+                # sparse parcl_ids can result in no data found.
+                # The post request can handle all 10k in one request, however the get request is one by one.
+                # get handles this direclty per parcl_id, while post handles all at once.
+                continue
+
+        return self._as_pd_dataframe(data_container)
 
     def sanitize_output(self, data: Dict[str, Any]) -> Dict[str, Any]:
-        """
-        Removes unwanted keys from the output data.
-        """
-        for key in DELETE_FROM_OUTPUT:
-            if key in data:
-                del data[key]
-        return data
+        return {k: v for k, v in data.items() if k not in DELETE_FROM_OUTPUT}
 
-    def _sync_request(
-        self,
-        parcl_id: int = None,
-        url: str = None,
-        params: Optional[Mapping[str, Any]] = None,
-        is_next: bool = False,
-        method: str = "GET",
-    ) -> Any:
-        if url:
-            url = url
-        elif parcl_id:
-            url = self.url.format(parcl_id=parcl_id)
-        else:
-            url = self.url
-        if method == "GET":
-            return self.get(url=url, params=params, is_next=is_next)
-        elif method == "POST":
-            return self.post(url=url, params=params)
-        else:
-            raise ValueError(f"Unsupported HTTP method: {method}")
+    def _as_pd_dataframe(self, data: List[Mapping[str, Any]]) -> pd.DataFrame:
+        data_container = deque()
+        for results in data:
+            if results is None:
+                continue
+            results = self.sanitize_output(results)
+            meta_fields = [k for k in results.keys() if k != "items"]
+            df = pd.json_normalize(results, record_path="items", meta=meta_fields)
+            updated_cols_names = [c.replace(".", "_") for c in df.columns.tolist()]
+            df.columns = updated_cols_names
+            data_container.append(df)
+
+        return safe_concat_and_format_dtypes(data_container)
 
     def error_handling(self, response: requests.Response) -> None:
         try:
-            error = ""
             error_details = response.json()
             error_message = error_details.get("detail", "No detail provided by API")
-            error = error_message
 
             if response.status_code == 403:
-                error = f"{error_message}. Visit https://dashboard.parcllabs.com for more information or reach out to team@parcllabs.com."
+                error_message += " Visit https://dashboard.parcllabs.com for more information or reach out to team@parcllabs.com."
+            elif response.status_code == 422:
+                error_message += error_details.get("msg", "validation error")
```

Comment:
this needs to be `error_details["details"]["msg"]`

Reply:
Really good catch. Updated it.